File: app/auth/modules/api_key_header.py
from __future__ import annotations

import os
from typing import Any

import httpx
from redis.asyncio import Redis

from app.auth.base import AuthContext, AuthResult, _invalidate_on_from_config


class ApiKeyHeaderAuth:
    """API key injected as a request header."""

    name = "api_key_header"

    def __init__(
        self,
        token: str,
        header: str,
        prefix: str,
        invalidate_on: set[int],
    ) -> None:
        self._token = token
        self._header = header
        self._prefix = prefix
        self._invalidate_on = invalidate_on

    @classmethod
    def from_config(
        cls, config: dict[str, Any], *, redis: Redis, http: httpx.AsyncClient
    ) -> ApiKeyHeaderAuth:
        env_name = config["env"]
        token = os.environ.get(env_name)
        if not token:
            raise ValueError(f"ApiKeyHeaderAuth: env var {env_name!r} is not set or empty.")
        return cls(
            token=token,
            header=config["header"],
            prefix=config.get("prefix", ""),
            invalidate_on=_invalidate_on_from_config(config),
        )

    async def apply(self, ctx: AuthContext) -> AuthResult:
        return AuthResult(headers={self._header: f"{self._prefix}{self._token}"})

    async def invalidate(self) -> None:
        pass

    def is_rejection(self, status_code: int, body: dict | None) -> bool:
        return status_code in self._invalidate_on

File: app/ts/src/auth/modules/api_key_header.ts
import type { AuthContext, AuthResult } from "../../core/types.ts";
import type { AuthModule, AuthDeps } from "../types.ts";
import { invalidateOnFromConfig } from "../types.ts";

export class ApiKeyHeaderAuth implements AuthModule {
  static readonly typeName = "api_key_header";

  private readonly token: string;
  private readonly header: string;
  private readonly prefix: string;
  private readonly invalidateOn: Set<number>;

  private constructor(
    token: string,
    header: string,
    prefix: string,
    invalidateOn: Set<number>,
  ) {
    this.token = token;
    this.header = header;
    this.prefix = prefix;
    this.invalidateOn = invalidateOn;
  }

  static fromConfig(
    config: Record<string, unknown>,
    deps: AuthDeps,
  ): ApiKeyHeaderAuth {
    if (typeof config["env"] !== "string") {
      throw new Error("ApiKeyHeaderAuth: 'env' config field is required.");
    }
    const envName = config["env"];
    const envVal = deps.env[envName];
    if (typeof envVal !== "string" || !envVal) {
      throw new Error(
        `ApiKeyHeaderAuth: env var '${envName}' is not set or empty.`,
      );
    }

    if (typeof config["header"] !== "string") {
      throw new Error("ApiKeyHeaderAuth: 'header' config field is required.");
    }

    const prefix = typeof config["prefix"] === "string" ? config["prefix"] : "";
    const invalidateOn = invalidateOnFromConfig(config, [401, 403]);

    return new ApiKeyHeaderAuth(envVal, config["header"], prefix, invalidateOn);
  }

  async apply(_ctx: AuthContext): Promise<AuthResult> {
    return { headers: { [this.header]: `${this.prefix}${this.token}` } };
  }

  async invalidate(): Promise<void> {}

  isRejection(statusCode: number, _body: unknown): boolean {
    return this.invalidateOn.has(statusCode);
  }
}
